# Stage 2. Geocoding Bangkok Neighbourhoods

In [1]:
from geopy import geocoders
geocoders.options.default_format_string = None
geocoders.options.default_user_agent = 'bkk_explorer'
import folium
from folium.features import DivIcon

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print('>>> Libraries likely imported')

>>> Libraries likely imported


In [2]:
# reload data from csv
bkk_khwaengs = pd.read_csv('csv/khwaengs.csv', dtype = 'str')
bkk_khwaengs

DCode              District       DistrictThai NCode  \
0      01           Phra Nakhon             พระนคร    01   
1      01           Phra Nakhon             พระนคร    02   
2      01           Phra Nakhon             พระนคร    03   
3      01           Phra Nakhon             พระนคร    04   
4      01           Phra Nakhon             พระนคร    05   
5      01           Phra Nakhon             พระนคร    06   
6      01           Phra Nakhon             พระนคร    07   
7      01           Phra Nakhon             พระนคร    08   
8      01           Phra Nakhon             พระนคร    09   
9      01           Phra Nakhon             พระนคร    10   
10     01           Phra Nakhon             พระนคร    11   
11     01           Phra Nakhon             พระนคร    12   
12     02                 Dusit              ดุสิต    01   
13     02                 Dusit              ดุสิต    02   
14     02                 Dusit              ดุสิต    03   
15     02                 Dusit              ดุสิต    04   
16     02                 Dusit              ดุสิต    06   
17     03             Nong Chok            หนองจอก    01   
18     03             Nong Chok            หนองจอก    02   
19     03             Nong Chok            หนองจอก    03   
20     03             Nong Chok            หนองจอก    04   
21     03             Nong Chok            หนองจอก    05   
22     03             Nong Chok            หนองจอก    06   
23     03             Nong Chok            หนองจอก    07   
24     03             Nong Chok            หนองจอก    08   
25     04              Bang Rak             บางรัก    01   
26     04              Bang Rak             บางรัก    02   
27     04              Bang Rak             บางรัก    03   
28     04              Bang Rak             บางรัก    04   
29     04              Bang Rak             บางรัก    05   
30     05             Bang Khen             บางเขน    02   
31     05             Bang Khen             บางเขน    08   
32     06             Bang Kapi            บางกะปิ    01   
33     06             Bang Kapi            บางกะปิ    08   
34     07            Pathum Wan            ปทุมวัน    01   
35     07            Pathum Wan            ปทุมวัน    02   
36     07            Pathum Wan            ปทุมวัน    03   
37     07            Pathum Wan            ปทุมวัน    04   
38     08  Pom Prap Sattru Phai  ป้อมปราบศัตรูพ่าย    01   
39     08  Pom Prap Sattru Phai  ป้อมปราบศัตรูพ่าย    02   
40     08  Pom Prap Sattru Phai  ป้อมปราบศัตรูพ่าย    03   
41     08  Pom Prap Sattru Phai  ป้อมปราบศัตรูพ่าย    04   
42     08  Pom Prap Sattru Phai  ป้อมปราบศัตรูพ่าย    05   
43     09          Phra Khanong            พระโขนง    05   
44     09          Phra Khanong            พระโขนง    10   
45     10              Min Buri            มีนบุรี    01   
46     10              Min Buri            มีนบุรี    02   
47     11           Lat Krabang          ลาดกระบัง    01   
48     11           Lat Krabang          ลาดกระบัง    02   
49     11           Lat Krabang          ลาดกระบัง    03   
50     11           Lat Krabang          ลาดกระบัง    04   
51     11           Lat Krabang          ลาดกระบัง    05   
52     11           Lat Krabang          ลาดกระบัง    06   
53     12              Yan Nawa            ยานนาวา    03   
54     12              Yan Nawa            ยานนาวา    04   
55     13        Samphanthawong        สัมพันธวงศ์    01   
56     13        Samphanthawong        สัมพันธวงศ์    02   
57     13        Samphanthawong        สัมพันธวงศ์    03   
58     14            Phaya Thai              พญาไท    01   
59     14            Phaya Thai              พญาไท    06   
60     15             Thon Buri             ธนบุรี    01   
61     15             Thon Buri             ธนบุรี    02   
62     15             Thon Buri             ธนบุรี    03   
63     15             Thon Buri             ธนบุรี    04   
64     15             Thon Buri             ธนบุรี    05   
65     15             Thon Buri             

## 1. Trying to get khwaeng coordinates with Geopy geocoders

### 2.1. Trying all available geocoders from the Geopy package to decide which one to use for the research

Getting the list of geocoders:

In [3]:
coders = dir(geocoders)
coders

['AlgoliaPlaces',
 'ArcGIS',
 'AzureMaps',
 'BANFrance',
 'Baidu',
 'BaiduV3',
 'Bing',
 'DataBC',
 'GeoNames',
 'GeocodeEarth',
 'GeocoderNotFound',
 'Geocodio',
 'Geolake',
 'GoogleV3',
 'Here',
 'HereV7',
 'IGNFrance',
 'LiveAddress',
 'MapBox',
 'MapQuest',
 'MapTiler',
 'Nominatim',
 'OpenCage',
 'OpenMapQuest',
 'Pelias',
 'Photon',
 'PickPoint',
 'SERVICE_TO_GEOCODER',
 'TomTom',
 'What3Words',
 'What3WordsV3',
 'Yandex',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 'algolia',
 'arcgis',
 'azure',
 'baidu',
 'banfrance',
 'base',
 'bing',
 'databc',
 'geocodeearth',
 'geocodio',
 'geolake',
 'geonames',
 'get_geocoder_for_service',
 'google',
 'here',
 'ignfrance',
 'mapbox',
 'mapquest',
 'maptiler',
 'nominatim',
 'opencage',
 'openmapquest',
 'options',
 'pelias',
 'photon',
 'pickpoint',
 'smartystreets',
 'tomtom',
 'what3words',
 'yandex']

Removing unnecessary items from the list

In [4]:
coders = coders[:coders.index("Yandex")+1]  # drop everything beyond Yandex
coders

['AlgoliaPlaces',
 'ArcGIS',
 'AzureMaps',
 'BANFrance',
 'Baidu',
 'BaiduV3',
 'Bing',
 'DataBC',
 'GeoNames',
 'GeocodeEarth',
 'GeocoderNotFound',
 'Geocodio',
 'Geolake',
 'GoogleV3',
 'Here',
 'HereV7',
 'IGNFrance',
 'LiveAddress',
 'MapBox',
 'MapQuest',
 'MapTiler',
 'Nominatim',
 'OpenCage',
 'OpenMapQuest',
 'Pelias',
 'Photon',
 'PickPoint',
 'SERVICE_TO_GEOCODER',
 'TomTom',
 'What3Words',
 'What3WordsV3',
 'Yandex']

In [5]:
def clean_list(list_name, junk):
    for item in junk:
        list_name.remove(item)

In [6]:
junk = ['GeocoderNotFound', 'SERVICE_TO_GEOCODER']
clean_list(coders, junk)

coders

['AlgoliaPlaces',
 'ArcGIS',
 'AzureMaps',
 'BANFrance',
 'Baidu',
 'BaiduV3',
 'Bing',
 'DataBC',
 'GeoNames',
 'GeocodeEarth',
 'Geocodio',
 'Geolake',
 'GoogleV3',
 'Here',
 'HereV7',
 'IGNFrance',
 'LiveAddress',
 'MapBox',
 'MapQuest',
 'MapTiler',
 'Nominatim',
 'OpenCage',
 'OpenMapQuest',
 'Pelias',
 'Photon',
 'PickPoint',
 'TomTom',
 'What3Words',
 'What3WordsV3',
 'Yandex']

Getting each geocoder try and find one of the more remote neighbourhoods - Khok Faet

In [7]:
geocoders.options.default_format_string = None

In [21]:
successful_coders = {}
address = 'แขวงโคกแฝด'
for coder in coders:
    print('Trying', coder)
    try:
        geolocator = geocoders.get_geocoder_for_service(coder)
    except: 
        print('Cannot get', coder, 'for service')
        print()
        continue
    try:
        print(geolocator)
        print('Searching:', address)
        location = geolocator().geocode(address)
        print('Found:', location, location.latitude, location.longitude)
        successful_coders[coder] = {'location': location, 'locator': geolocator}
    except:
        print(coder, 'cannot find the location')
        print()
        continue
    print()

Trying AlgoliaPlaces
Cannot get AlgoliaPlaces for service

Trying ArcGIS
<class 'geopy.geocoders.arcgis.ArcGIS'>
Searching: แขวงโคกแฝด
Found: เขตหนองจอก กุรงเทพมหานคร 13.830434508059682 100.82415605758723

Trying AzureMaps
Cannot get AzureMaps for service

Trying BANFrance
<class 'geopy.geocoders.banfrance.BANFrance'>
Searching: แขวงโคกแฝด
BANFrance cannot find the location

Trying Baidu
<class 'geopy.geocoders.baidu.Baidu'>
Searching: แขวงโคกแฝด
Baidu cannot find the location

Trying BaiduV3
<class 'geopy.geocoders.baidu.BaiduV3'>
Searching: แขวงโคกแฝด
BaiduV3 cannot find the location

Trying Bing
<class 'geopy.geocoders.bing.Bing'>
Searching: แขวงโคกแฝด
Bing cannot find the location

Trying DataBC
<class 'geopy.geocoders.databc.DataBC'>
Searching: แขวงโคกแฝด
Found: BC 53.913051 -122.7452849

Trying GeoNames
<class 'geopy.geocoders.geonames.GeoNames'>
Searching: แขวงโคกแฝด
GeoNames cannot find the location

Trying GeocodeEarth
<class 'geopy.geocoders.geocodeearth.GeocodeEarth'>
Search

##### The only geocoders that can find **โคกแฝด** neighbourhood in Bangkok with no intricate setup procedure are:

In [22]:
for k, v in successful_coders.items():
    print(k, ":")
    print('\t', v)

ArcGIS :
	 {'location': Location(เขตหนองจอก กุรงเทพมหานคร, (13.830434508059682, 100.82415605758723, 0.0)), 'locator': <class 'geopy.geocoders.arcgis.ArcGIS'>}
DataBC :
	 {'location': Location(BC, (53.913051, -122.7452849, 0.0)), 'locator': <class 'geopy.geocoders.databc.DataBC'>}
Nominatim :
	 {'location': Location(แขวงโคกแฝด, เขตหนองจอก, กรุงเทพมหานคร, 10530, ประเทศไทย, (13.8537368, 100.828743, 0.0)), 'locator': <class 'geopy.geocoders.nominatim.Nominatim'>}


and the location returned by DataBC is definitely way off with longitude value of -122.7452849.

In [23]:
successful_coders.pop('DataBC')

{'location': Location(BC, (53.913051, -122.7452849, 0.0)),
 'locator': geopy.geocoders.databc.DataBC}

Visualizing the other two results on map:

In [24]:
center_latitude = sum(value['location'].latitude for value in successful_coders.values())/len(successful_coders)
center_longitude = sum(value['location'].longitude for value in successful_coders.values())/len(successful_coders)

center_latitude, center_longitude

(13.842085654029841, 100.82644952879362)

In [25]:
map_khokfaet = folium.Map(location=[center_latitude, center_longitude], zoom_start=13)

for coder, value in successful_coders.items():
    location = (value['location'].latitude, value['location'].longitude)
    # add marker to map
    folium.CircleMarker(
            location,
            radius=5,
            color='blue',
            fill=True,
            fill_color='#3186cc',
            fill_opacity=0.7,
            parse_html=False).add_to(map_khokfaet)  
    folium.map.Marker(
            location,
            icon=DivIcon(
                # icon_size=(30,30),
                icon_anchor=(-2,-2),
                html=f'<div style="font-size: 14pt">{coder}</div>')).add_to(map_khokfaet)

map_khokfaet

### 2.2. Using ArcGIS geocoder to get coordinates for each of 180 neighbourhoods

In [26]:
geolocator =  geocoders.get_geocoder_for_service('ArcGIS')

print(geolocator())

#### 2.1.1. Trying on just one khwaeng first

#### 2.1.2. Geocoding all neighbourhoods

In [27]:
bkk_khwaengs.columns

Index(['DCode', 'District', 'DistrictThai', 'NCode', 'Neighbourhood',
       'NeighbourhoodThai', 'Latitude', 'Longitude'],
      dtype='object')

In [28]:
lats = list() # creating an empty list to store lattitudes
longs = list() # and longitudes
for i, row in bkk_khwaengs.iterrows() :
    location = geolocator.geocode(row['NeighbourhoodThai'])
    lats.append(location.latitude)
    longs.append(location.longitude)
    print(i, location, location.latitude, location.longitude)

TypeError: geocode() missing 1 required positional argument: 'query'

and adding each neighbourhood coordinates to the dataframe in respective columns:

In [ ]:
bkk_khwaengs.columns

In [ ]:
bkk_khwaengs['Latitude'] = lats
bkk_khwaengs['Longitude'] = longs
bkk_khwaengs

In [ ]:
bkk_khwaengs.describe(include='all')

In [ ]:
bkk_khwaengs.dtypes

In [ ]:
bkk_khwaengs.to_csv('khwaengs_geocoded2.csv', index = False)

#### 2.1.3. Visualizing bangkok neighbourhoods on Folium map

In [ ]:
# create map of Bangkok using latitude and longitude values
latitude = bkk_khwaengs.Latitude.mean() + 0.01
longitude = bkk_khwaengs.Longitude.mean() + 0.06
map_bangkok = folium.Map(location = (latitude, longitude), width = 900, zoom_start=11)

# add markers to map
for lat, lng, neighbourhood in zip(bkk_khwaengs['Latitude'], bkk_khwaengs['Longitude'], bkk_khwaengs['Neighbourhood']):
    label = '{}'.format(neighbourhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_bangkok)  
    
map_bangkok